File: Ad-hoc/SST.m
function [x] = SST(A, b)
    % if the matrix is not upper triangular, we cannot solve the system
    % tril(A, -1) gets us what is strictly under the main diagonal
    % max(matrix) gets us a list of every column's max
    % max(list) gets us the maximum from the list
    % [NOTE] There is a function called "istriu(A)" which returns 1 if A is an
    %   upper triangular matrix but it doesn't catch the e-16 errors
    %   which should be ignored
    if max(max(abs(tril(A, -1)))) > 10^3 * eps
        disp('The matrix A is not upper triangular!');
        x = NaN;
        return;
    endif

    n = length(b);
    x = zeros(n, 1);

    % calculate xn
    x(n) = b(n) / A(n, n);

    % calculate x(i) backwards
    for i = (n - 1): -1 : 1
        sum_of_xs = A(i, (i + 1) : n) * x((i + 1) : n, 1);
        x(i) = (b(i) - sum_of_xs) / A(i, i);
    endfor
endfunction

File: python/ad-hoc/SST.py
import  numpy as np
def SST(A, b):
  # if the matrix is not upper triangular, we cannot solve the system
  if not np.allclose(A, np.triu(A)):
    print("The matrix A is not upper triangular!")
    return np.nan

  n = len(b)
  x = np.zeros((n, 1))

  # calculate xn
  x[n - 1, 0] = b[n - 1, 0] / A[n - 1, n - 1]

  # calculate x(i) backwards
  for i in range(n - 2, -1, -1):
    sum_of_xs = np.dot(A[i, (i + 1) : n], x[(i + 1) : n, 0])
    x[i, 0] = (b[i, 0] - sum_of_xs) / A[i, i]
    
  return x